### DISCUSSION TITLE
Example on how to read from sdcard forth code and execute it

### DISCUSSION THREAD

**Discussion by manes6969** [tabemann/zeptoforth#42] [Q&A]
Hi ,

I managed , following the wiki pages to create and read on sdcard (low level so far).
I also read about redirecting. But unfortunately there are no examples on how to redirect for example a file on the sdcard to the serial console. At least I think this is the way to do it. Is there any info somewhere on how to do this ?
BTW : zeptoforth is AWESOME !!!
Regards,
Ronny

**Comment by tabemann** [comment]
It is mentioned in [Console Redirection in zeptoforth](https://github.com/tabemann/zeptoforth/wiki/Console-Redirection-in-zeptoforth) but I never got around to writing any code examples in that wiki page. I should probably write some examples now.

Thanks for the props, BTW :-)

**Comment by tabemann** [comment]
@manes6969 BTW, I have now added some code examples to that wiki page. I hope they are helpful.

**Comment by manes6969** [comment]
Hi tabemann,

Thanks very much for the code.!  It was certainly helpfull , but I still struggle to 'execute' a forth program from sdcard.
I ran this code 
begin-module fat32-test  ok
  ok
  oo import  ok
  fat32 import  ok
  simple-fat32 import  ok
    ok
  <simple-fat32-fs> class-size buffer: my-fs  ok
    ok
  0 constant my-spi  ok
  2 constant spi0-sck-pin  ok
  3 constant spi0-tx-pin  ok
  4 constant spi0-rx-pin  ok
  5 constant cs-pin  ok
    ok
  : run-test ( -- )  ok
    spi0-sck-pin spi0-tx-pin spi0-rx-pin cs-pin my-spi <simple-fat32-fs> my-fs init-object  ok
    false my-fs write-through!  ok
    my-fs fat32-tools::current-fs!  ok
    s" /test.fs" [:  ok
      begin key? while key emit repeat  ok
    ;] fat32-tools::with-file-input  ok
  ;  ok
    ok
end-module  ok
fat32-test::run-test : lll 10 0 do i . loop ;

                                              ok
The **: lll 10 0 do i . loop ;** are the contents of the file test.fs that I've written on the sdcard on my pc.
So it is displayed , and the ok suggests that lll is compiled , but when I try lll , an error is displayed
**lll unable to parse: lll**
even if I press enter after the last line.
Do you have a clue what could be wrong here ?
Kind regards ,
Ronny

**Comment by manes6969** [comment]
Hi tabemann ,

I think I know the problem : the bytes from the sdcard are just printed out , there not copied into the evaluation buffer. correct. ?
Trying to figure out how to do this ...
Regards,
Ronny

**Reply by tabemann** [reply]
The solution is a follows (forgive my multiple incomplete messages that I accidentally sent and then deleted):

```
begin-module fat32-test

  oo import
  fat32 import
  simple-fat32 import
  
  <simple-fat32-fs> class-size buffer: my-fs
  
  0 constant my-spi
  2 constant spi0-sck-pin
  3 constant spi0-tx-pin
  4 constant spi0-rx-pin
  5 constant cs-pin
  
  : run-test ( -- )
    spi0-sck-pin spi0-tx-pin spi0-rx-pin cs-pin my-spi <simple-fat32-fs> my-fs init-object
    false my-fs write-through!
    my-fs fat32-tools::current-fs!
    s" /test.fs" fat32-tools::included
  ;
  
end-module
```

When you execute `fat32-test::run-test` it should execute the contents of `/test.fs`.

**Reply by tabemann** [reply]
And yes, redirecting input from a file  just redirects the source of `key` and `key?` - it does not by itself direct the file into the evaluation buffer.

**Comment by manes6969** [comment]
Hi tabemann ,
Thanks for the help , I will close this discussion
Regards,
Ronny
